src/rust/lib_ccxr/src/util/encoding.rs

```diff
@@ -0,0 +1,811 @@
+    /// let s: EncodedString = Line21String::from_vec(v).into();
+    /// assert_eq!(s.encode_to(Encoding::Utf8), "résumé".to_string().into())
+    /// ```
+    pub fn encode_to(&self, encoding: Encoding) -> EncodedString {
+        match encoding {
+            Encoding::Line21 => EncodedString::Line21(self.to_line21()),
+            Encoding::Latin1 => EncodedString::Latin1(self.to_latin1()),
+            Encoding::Ucs2 => EncodedString::Ucs2(self.to_ucs2()),
+            Encoding::Utf8 => EncodedString::Utf8(self.to_utf8()),
+        }
+    }
+
+    /// Converts the [`EncodedString`] to lowercase, returning a new [`EncodedString`].
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use lib_ccxr::util::encoding::*;
+    /// let a = vec![0x72, 0x5c, 0x73, 0x75, 0x6d, 0x5c]; // résumé in Line 21 encoding
+    /// let b = vec![0x72, 0x91, 0x73, 0x75, 0x6d, 0x91]; // RÉSUMÉ in Line 21 encoding
+    /// let sa: EncodedString = Line21String::from_vec(a).into();
+    /// let sb: EncodedString = Line21String::from_vec(b).into();
+    /// assert_eq!(sb.to_lowercase(), sa)
+    /// ```
+    pub fn to_lowercase(&self) -> EncodedString {
+        match self {
+            EncodedString::Line21(l) => l.to_lowercase().into(),
+            EncodedString::Latin1(l1) => l1.to_lowercase().into(),
+            EncodedString::Ucs2(u) => u.to_lowercase().into(),
+            EncodedString::Utf8(s) => s.to_lowercase().into(),
+        }
+    }
+
+    /// Converts the [`EncodedString`] to uppercase, returning a new [`EncodedString`].
+    ///
+    /// # Examples
+    /// ```rust
+    /// # use lib_ccxr::util::encoding::*;
+    /// let a = vec![0x72, 0x5c, 0x73, 0x75, 0x6d, 0x5c]; // résumé in Line 21 encoding
+    /// let b = vec![0x52, 0x91, 0x53, 0x55, 0x4d, 0x91]; // RÉSUMÉ in Line 21 encoding
+    /// let sa: EncodedString = Line21String::from_vec(a).into();
+    /// let sb: EncodedString = Line21String::from_vec(b).into();
+    /// assert_eq!(sa.to_uppercase(), sb)
+    /// ```
+    pub fn to_uppercase(&self) -> EncodedString {
+        match self {
+            EncodedString::Line21(l) => l.to_uppercase().into(),
+            EncodedString::Latin1(l1) => l1.to_uppercase().into(),
+            EncodedString::Ucs2(u) => u.to_uppercase().into(),
+            EncodedString::Utf8(s) => s.to_uppercase().into(),
+        }
+    }
+}
+
+impl From<Line21String> for EncodedString {
+    fn from(value: Line21String) -> Self {
+        EncodedString::Line21(value)
+    }
+}
+
+impl From<Latin1String> for EncodedString {
+    fn from(value: Latin1String) -> Self {
+        EncodedString::Latin1(value)
+    }
+}
+
+impl From<Ucs2String> for EncodedString {
+    fn from(value: Ucs2String) -> Self {
+        EncodedString::Ucs2(value)
+    }
+}
+
+impl From<String> for EncodedString {
+    fn from(value: String) -> Self {
+        EncodedString::Utf8(value)
+    }
+}
+
+fn latin1_to_line21(_c: Latin1Char) -> Line21Char {
+    todo!()
+}
```

Comment:
1 `todo` remaining
Researching on this one
@PunitLodha @elbertronnie Any idea about it, This function is not included in C but rust match(line 417) needed it.

Reply:
Even though it is not required in C, I wanted all forms to be convertable to all other forms of String simply because of completeness. Implementing this would take some time which is why I left a todo in there to do it later.

If you want to implement it you can do it by inverting the logic in `line21_to_latin1`.

Reply:
> Even though it is not required in C, I wanted all forms to be convertable to all other forms of String simply because of completeness. Implementing this would take some time which is why I left a todo in there to do it later.

Yeah exactly @elbertronnie , I removed some todo from your code, just left with this one only

> If you want to implement it you can do it by inverting the logic in line21_to_latin1.

I am trying to do that only